File: Shared/CommandQueue.swift
import Foundation

struct CommandQueue {
    private var pending: [Command] = []
    var isEmpty: Bool { pending.isEmpty }
    var count: Int { pending.count }

    @discardableResult
    mutating func append(_ command: Command) -> Bool {
        guard writeCount(command) != nil else { removeAll(); return false }
        if case .move(let dx, let dy) = command,
           case .move(let oldX, let oldY) = pending.last,
           (oldX + dx).isFinite, (oldY + dy).isFinite {
            pending[pending.count - 1] = .move(dx: oldX + dx, dy: oldY + dy)
        } else if case .scroll(let dx, let dy) = command,
                  case .scroll(let oldX, let oldY) = pending.last,
                  (oldX + dx).isFinite, (oldY + dy).isFinite {
            pending[pending.count - 1] = .scroll(dx: oldX + dx, dy: oldY + dy)
        } else {
            pending.append(command)
        }
        var writes = 0
        for command in pending {
            guard let count = writeCount(command) else { removeAll(); return false }
            writes += count
        }
        guard writes <= 64 else { removeAll(); return false }
        return true
    }

    private func writeCount(_ command: Command) -> Int? {
        switch command {
        case .move(let x, let y), .scroll(let x, let y):
            guard x.isFinite, y.isFinite, abs(x) <= 25_600, abs(y) <= 25_600 else { return nil }
            return max(1, Int(ceil(max(abs(x), abs(y)) / 400)))
        default: return 1
        }
    }

    mutating func popFirst() -> Command? {
        guard !pending.isEmpty else { return nil }
        var command = pending.removeFirst()
        switch command {
        case .move(let x, let y), .scroll(let x, let y):
            let dx = min(400, max(-400, x))
            let dy = min(400, max(-400, y))
            if case .move = command {
                command = .move(dx: dx, dy: dy)
                if x != dx || y != dy { pending.insert(.move(dx: x - dx, dy: y - dy), at: 0) }
            } else {
                command = .scroll(dx: dx, dy: dy)
                if x != dx || y != dy { pending.insert(.scroll(dx: x - dx, dy: y - dy), at: 0) }
            }
        default: break
        }
        return command
    }

    mutating func removeAll() { pending.removeAll() }
}

File: android/protocol/src/main/kotlin/systems/edmundlim/stagewand/protocol/CommandQueue.kt
package systems.edmundlim.stagewand.protocol

import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

class CommandQueue {
    private val pending = ArrayDeque<Command>()
    val isEmpty: Boolean get() = pending.isEmpty()
    val count: Int get() = pending.size

    fun append(command: Command): Boolean {
        if (writeCount(command) == null) {
            removeAll()
            return false
        }
        val last = pending.lastOrNull()
        if (command is Command.Move && last is Command.Move) {
            val x = last.dx + command.dx
            val y = last.dy + command.dy
            if (x.isFinite() && y.isFinite()) {
                pending.removeLast()
                pending.addLast(Command.Move(x, y))
                return checkCapacity()
            }
        } else if (command is Command.Scroll && last is Command.Scroll) {
            val x = last.dx + command.dx
            val y = last.dy + command.dy
            if (x.isFinite() && y.isFinite()) {
                pending.removeLast()
                pending.addLast(Command.Scroll(x, y))
                return checkCapacity()
            }
        }
        pending.addLast(command)
        return checkCapacity()
    }

    private fun checkCapacity(): Boolean {
        var writes = 0
        for (command in pending) {
            val count = writeCount(command)
            if (count == null || writes + count > 64) {
                removeAll()
                return false
            }
            writes += count
        }
        return true
    }

    private fun writeCount(command: Command): Int? {
        val delta = when (command) {
            is Command.Move -> command.dx to command.dy
            is Command.Scroll -> command.dx to command.dy
            else -> return 1
        }
        val (x, y) = delta
        if (!x.isFinite() || !y.isFinite() || abs(x) > 25_600 || abs(y) > 25_600) return null
        return max(1, ceil(max(abs(x), abs(y)) / 400).toInt())
    }

    fun popFirst(): Command? {
        if (pending.isEmpty()) return null
        val command = pending.removeFirst()
        if (command is Command.Move) {
            val dx = min(400.0, max(-400.0, command.dx))
            val dy = min(400.0, max(-400.0, command.dy))
            if (command.dx != dx || command.dy != dy) {
                pending.addFirst(Command.Move(command.dx - dx, command.dy - dy))
            }
            return Command.Move(dx, dy)
        }
        if (command is Command.Scroll) {
            val dx = min(400.0, max(-400.0, command.dx))
            val dy = min(400.0, max(-400.0, command.dy))
            if (command.dx != dx || command.dy != dy) {
                pending.addFirst(Command.Scroll(command.dx - dx, command.dy - dy))
            }
            return Command.Scroll(dx, dy)
        }
        return command
    }

    fun removeAll() {
        pending.clear()
    }
}
